shared/modules/typed-signature.ts

```diff
@@ -0,0 +1,39 @@
+import { MessageTypes } from '@metamask/eth-sig-util';
+
+/**
+ * Recursively sanitizes a message object based on the provided types and primary type.
+ *
+ * @param message - The message object to sanitize.
+ * @param types - The types defined in the typed message.
+ * @param primaryType - The primary type of the message.
+ * @returns The sanitized message object.
+ */
+export function sanitizeMessageRecursively<T extends MessageTypes>(
+  message: Record<string, unknown>,
+  types: T,
+  primaryType: string | number,
+): Record<string, unknown> {
+  const sanitizedMessage: Record<string, unknown> = {};
+  const typeDefinition = types[primaryType];
+
+  if (!typeDefinition) {
+    return message;
+  }
+
+  for (const field of typeDefinition) {
+    const { name, type } = field;
+    if (message[name] !== undefined) {
+      if (types[type]) {
```

Comment:
I understood the normalization we wanted to do in the middleware wasn't what the UI did, but what `eth-sig-util` did internally when it cast the values based on the type.

This would ensure what Blockaid received would exactly match what we ultimately signed.